stage2: compiler crash calling fn with wrong number of args

### Zig Version

0.10.0-dev.1606+97a53bb8a

### Steps to Reproduce

Compile this code with stage2:

```zig
fn foo(x: u32) u32 {
    return x;
}

pub fn main() void {
    foo(1, 1);
}
```

### Expected Behavior

stage1 gives the following compiler error:

```
$ zig build-exe param_type_bug.zig
./param_type_bug.zig:6:8: error: expected 1 argument(s), found 2
    foo(1, 1);
       ^
./param_type_bug.zig:1:1: note: declared here
fn foo(x: u32) u32 {
```

### Actual Behavior

```
$ zig build-exe param_type_bug.zig
thread 9301805 panic: reached unreachable code
Analyzing param_type_bug.zig: param_type_bug.zig:main
      %15 = extended(dbg_block_begin())
      %16 = dbg_stmt(2, 5)
      %17 = decl_val("foo") token_offset:6:5
      %18 = dbg_stmt(2, 8)
      %19 = param_type(%17, 0)
    > %20 = param_type(%17, 1)
      %22 = extended(dbg_block_end())
      %21 = call(nodiscard .auto, %17, [@Zir.Inst.Ref.one, @Zir.Inst.Ref.one]) node_offset:6:8
      %23 = ret_tok(@Zir.Inst.Ref.void_value) token_offset:7:1
    For full context, use the command
      zig ast-check -t param_type_bug.zig


/Users/<USER>/repos/zig/build/lib/zig/std/debug.zig:234:14: 0x10b327d58 in std.debug.assert (zig)
    if (!ok) unreachable; // assertion failure
             ^
/Users/<USER>/repos/zig/src/Sema.zig:3664:15: 0x10ba05c8b in Sema.zirParamType (zig)
        assert(fn_info.is_var_args);
              ^
/Users/<USER>/repos/zig/src/Sema.zig:741:67: 0x10b88fdb1 in Sema.analyzeBodyInner (zig)
            .param_type                   => try sema.zirParamType(block, inst),
                                                                  ^
/Users/<USER>/repos/zig/src/Sema.zig:568:30: 0x10b88112b in Sema.analyzeBody (zig)
    _ = sema.analyzeBodyInner(block, body) catch |err| switch (err) {
                             ^
/Users/<USER>/repos/zig/src/Module.zig:4820:21: 0x10b6cd3a7 in Module.analyzeFnBody (zig)
    sema.analyzeBody(&inner_block, fn_info.body) catch |err| switch (err) {
                    ^
/Users/<USER>/repos/zig/src/Module.zig:3603:40: 0x10b6aebb7 in Module.ensureFuncBodyAnalyzed (zig)
            var air = mod.analyzeFnBody(decl, func, sema_arena) catch |err| switch (err) {
                                       ^
/Users/<USER>/repos/zig/src/Compilation.zig:2745:42: 0x10b470b70 in Compilation.processOneJob (zig)
            module.ensureFuncBodyAnalyzed(func) catch |err| switch (err) {
                                         ^
/Users/<USER>/repos/zig/src/Compilation.zig:2691:30: 0x10b45d613 in Compilation.performAllTheWork (zig)
            try processOneJob(self, work_item, main_progress_node);
                             ^
/Users/<USER>/repos/zig/src/Compilation.zig:2094:31: 0x10b45675a in Compilation.update (zig)
    try comp.performAllTheWork();
                              ^
/Users/<USER>/repos/zig/src/main.zig:3101:20: 0x10b3bfd70 in updateModule (zig)
    try comp.update();
                   ^
/Users/<USER>/repos/zig/src/main.zig:2788:17: 0x10b34a10d in buildOutputType (zig)
    updateModule(gpa, comp, hook) catch |err| switch (err) {
                ^
/Users/<USER>/repos/zig/src/main.zig:210:31: 0x10b32faeb in mainArgs (zig)
        return buildOutputType(gpa, arena, args, .{ .build = .Exe });
                              ^
/Users/<USER>/repos/zig/src/main.zig:165:20: 0x10b32ecf2 in main (zig)
    return mainArgs(gpa, arena, args);
                   ^
/Users/<USER>/repos/zig/build/lib/zig/std/start.zig:574:37: 0x10b8392d8 in std.start.callMain (zig)
            const result = root.main() catch |err| {
                                    ^
/Users/<USER>/repos/zig/build/lib/zig/std/start.zig:508:12: 0x10b331607 in std.start.callMainWithArgs (zig)
    return @call(.{ .modifier = .always_inline }, callMain, .{});
           ^
/Users/<USER>/repos/zig/build/lib/zig/std/start.zig:473:12: 0x10b331545 in std.start.main (zig)
    return @call(.{ .modifier = .always_inline }, callMainWithArgs, .{ @intCast(usize, c_argc), c_argv, envp });
           ^
???:?:?: 0x7fff68290cc8 in ??? (???)
???:?:?: 0x2 in ??? (???)
zsh: abort      ~/repos/zig/stage2/bin/zig build-exe param_type_bug.zig
```

--- a/build.zig
+++ b/build.zig
@@ -613,7 +613,8 @@ fn addCxxKnownPath(
         ctx.cxx_compiler,
         b.fmt("-print-file-name={s}", .{objname}),
     });
-    const path_unpadded = mem.tokenize(u8, path_padded, "\r\n").next().?;
+    var tokenizer = mem.tokenize(u8, path_padded, "\r\n");
+    const path_unpadded = tokenizer.next().?;
     if (mem.eql(u8, path_unpadded, objname)) {
         if (errtxt) |msg| {
             std.debug.print("{s}", .{msg});

--- a/src/Sema.zig
+++ b/src/Sema.zig
@@ -3661,8 +3661,12 @@ fn zirParamType(sema: *Sema, block: *Block, inst: Zir.Inst.Index) CompileError!A
         fn_ty.fnInfo();
 
     if (param_index >= fn_info.param_types.len) {
-        assert(fn_info.is_var_args);
-        return sema.addType(Type.initTag(.var_args_param));
+        if (fn_info.is_var_args) {
+            return sema.addType(Type.initTag(.var_args_param));
+        }
+        // TODO implement begin_call/end_call Zir instructions and check
+        // argument count before casting arguments to parameter types.
+        return sema.fail(block, callee_src, "wrong number of arguments", .{});
     }
 
     if (fn_info.param_types[param_index].tag() == .generic_poison) {

--- a/src/Sema.zig
+++ b/src/Sema.zig
@@ -2317,8 +2317,8 @@ fn zirErrorSetDecl(
     const extra_index_end = extra_index + (extra.data.fields_len * 2);
     while (extra_index < extra_index_end) : (extra_index += 2) { // +2 to skip over doc_string
         const str_index = sema.code.extra[extra_index];
-        const name = try new_decl_arena_allocator.dupe(u8, sema.code.nullTerminatedString(str_index));
-        const result = names.getOrPutAssumeCapacity(name);
+        const kv = try sema.mod.getErrorValue(sema.code.nullTerminatedString(str_index));
+        const result = names.getOrPutAssumeCapacity(kv.key);
         assert(!result.found_existing); // verified in AstGen
     }
 
@@ -13113,11 +13113,10 @@ fn zirReify(sema: *Sema, block: *Block, inst: Zir.Inst.Index) CompileError!Air.I
                 // TODO use reflection instead of magic numbers here
                 // error_set: type,
                 const name_val = struct_val[0];
+                const name_str = try name_val.toAllocatedBytes(Type.initTag(.const_slice_u8), sema.arena, target);
 
-                names.putAssumeCapacityNoClobber(
-                    try name_val.toAllocatedBytes(Type.initTag(.const_slice_u8), sema.arena, target),
-                    {},
-                );
+                const kv = try sema.mod.getErrorValue(name_str);
+                names.putAssumeCapacityNoClobber(kv.key, {});
             }
 
             // names must be sorted

--- a/src/value.zig
+++ b/src/value.zig
@@ -461,7 +461,7 @@ pub const Value = extern union {
             => unreachable,
 
             .ty, .lazy_align => {
-                const payload = self.castTag(.ty).?;
+                const payload = self.cast(Payload.Ty).?;
                 const new_payload = try arena.create(Payload.Ty);
                 new_payload.* = .{
                     .base = payload.base,
@@ -718,7 +718,7 @@ pub const Value = extern union {
             .lazy_align => {
                 try out_stream.writeAll("@alignOf(");
                 try val.castTag(.lazy_align).?.data.dump("", options, out_stream);
-                try out_stream.writeAll(")");
+                return try out_stream.writeAll(")");
             },
             .int_type => {
                 const int_type = val.castTag(.int_type).?.data;
@@ -2477,6 +2477,13 @@ pub const Value = extern union {
             .bool_true,
             .the_only_possible_value,
             => return hashInt(ptr_val, hasher, target),
+
+            .lazy_align => {
+                // Bit weird to have this here but this function is also called
+                // on integers.
+                const ty = ptr_val.castTag(.lazy_align).?.data;
+                ty.hashWithHasher(hasher, target);
+            },
 
             else => unreachable,
         }

